File: tests/Wording.Core.Tests/WordPackReaderTests.cs
using System.Text;
using Wording.Core.Packs;

namespace Wording.Core.Tests;

public class WordPackReaderTests
{
    static byte[] Bytes(string json) => Encoding.UTF8.GetBytes(json);

    const string Valid = """
        {
          "id": "travel-basics",
          "name": "Travel basics",
          "description": "Everyday phrases",
          "words": [
            { "original": "airport", "translation": "aeropuerto" },
            { "original": "ticket", "translation": "billete" }
          ]
        }
        """;

    [Fact]
    public void Read_AcceptsAWellFormedPack()
    {
        var pack = WordPackReader.Read(Bytes(Valid));

        Assert.Equal("travel-basics", pack.Id);
        Assert.Equal("Travel basics", pack.Name);
        Assert.Equal("Everyday phrases", pack.Description);
        Assert.Equal(2, pack.Words.Count);
        Assert.Equal("aeropuerto", pack.Words[0].Translation);
    }

    [Fact]
    public void Read_KeepsNonAsciiIntact()
    {
        var pack = WordPackReader.Read(Bytes("""
            { "id": "pl", "name": "Polski", "words": [
              { "original": "default", "translation": "domyślnie" }] }
            """));

        Assert.Equal("domyślnie", pack.Words[0].Translation);
    }

    [Theory]
    [InlineData("not json at all")]
    [InlineData("{ \"id\": ")]
    [InlineData("[]")]
    public void Read_RefusesWhatIsNotAPack(string json)
    {
        Assert.Equal(PackProblem.Malformed, Problem(json));
    }

    [Fact]
    public void Read_RefusesAPackThatChoosesItsOwnFileName()
    {
        var json = Valid.Replace("travel-basics", "../../words", StringComparison.Ordinal);

        Assert.Equal(PackProblem.UnsafeId, Problem(json));
    }

    [Fact]
    public void Read_RefusesAPackWithNoName()
    {
        Assert.Equal(PackProblem.Malformed, Problem(Valid.Replace("Travel basics", "   ", StringComparison.Ordinal)));
    }

    [Fact]
    public void Read_RefusesAPayloadOverTheLimit()
    {
        // Padding inside the description, so the file stays valid JSON.
        var padding = new string('x', PackLimits.MaxPayloadBytes + 1);
        var json = Valid.Replace("Everyday phrases", padding, StringComparison.Ordinal);

        Assert.Equal(PackProblem.TooLarge, Problem(json));
    }

    [Fact]
    public void Read_RefusesTooManyWords()
    {
        var entries = string.Join(",", Enumerable
            .Range(0, PackLimits.MaxWords + 1)
            .Select(index => $$"""{ "original": "w{{index}}", "translation": "t{{index}}" }"""));

        Assert.Equal(
            PackProblem.TooLarge,
            Problem($$"""{ "id": "big", "name": "Big", "words": [{{entries}}] }"""));
    }

    [Fact]
    public void Read_RefusesAFieldTooLongToBeAWord()
    {
        var essay = new string('a', PackLimits.MaxFieldLength + 1);
        var json = $$"""{ "id": "x", "name": "X", "words": [{ "original": "{{essay}}", "translation": "t" }] }""";

        Assert.Equal(PackProblem.Malformed, Problem(json));
    }

    [Fact]
    public void Read_SkipsBlankEntriesButKeepsTheRest()
    {
        var json = """
            { "id": "x", "name": "X", "words": [
              { "original": "  ", "translation": "empty" },
              { "original": "keep", "translation": "" },
              { "original": "airport", "translation": "aeropuerto" }] }
            """;

        var pack = WordPackReader.Read(Bytes(json));

        Assert.Equal("airport", Assert.Single(pack.Words).Original);
    }

    [Fact]
    public void Read_RefusesAPackWhereNothingUsableIsLeft()
    {
        var json = """{ "id": "x", "name": "X", "words": [{ "original": " ", "translation": " " }] }""";

        Assert.Equal(PackProblem.Empty, Problem(json));
    }

    [Fact]
    public void Read_RefusesAPackWithNoWordsAtAll()
    {
        Assert.Equal(PackProblem.Empty, Problem("""{ "id": "x", "name": "X", "words": [] }"""));
    }

    [Fact]
    public void Read_FoldsControlCharactersThatWouldReachANotification()
    {
        var json = """
            { "id": "x", "name": "X", "words": [
              { "original": "two\nlines", "translation": "\ttabbed " }] }
            """;

        var word = Assert.Single(WordPackReader.Read(Bytes(json)).Words);

        Assert.Equal("two lines", word.Original);
        Assert.Equal("tabbed", word.Translation);
    }

    [Fact]
    public void Read_TruncatesTheNameRatherThanRefusingThePack()
    {
        // Display-only, and the reader cannot expect the user to fix someone else's file.
        var json = Valid.Replace("Travel basics", new string('n', 200), StringComparison.Ordinal);

        Assert.Equal(PackLimits.MaxNameLength, WordPackReader.Read(Bytes(json)).Name.Length);
    }

    static PackProblem Problem(string json) =>
        Assert.Throws<WordPackException>(() => WordPackReader.Read(Bytes(json))).Problem;
}

File: macos/Tests/WordingKitTests/WordPackReaderTests.swift
import Foundation
import Testing

@testable import WordingKit

@Suite struct WordPackReaderTests {
    static let valid = """
        {
          "id": "travel-basics",
          "name": "Travel basics",
          "description": "Everyday phrases",
          "words": [
            { "original": "airport", "translation": "aeropuerto" },
            { "original": "ticket", "translation": "billete" }
          ]
        }
        """

    static func data(_ json: String) -> Data { Data(json.utf8) }

    @Test func acceptsAWellFormedPack() throws {
        let pack = try WordPackReader.read(Self.data(Self.valid))

        #expect(pack.id == "travel-basics")
        #expect(pack.name == "Travel basics")
        #expect(pack.description == "Everyday phrases")
        #expect(pack.words.count == 2)
        #expect(pack.words[0].translation == "aeropuerto")
    }

    @Test func keepsNonAsciiIntact() throws {
        let pack = try WordPackReader.read(
            Self.data("""
                { "id": "pl", "name": "Polski", "words": [
                  { "original": "default", "translation": "domyślnie" }] }
                """))

        #expect(pack.words[0].translation == "domyślnie")
    }

    @Test(arguments: ["not json at all", "{ \"id\": ", "[]"])
    func refusesWhatIsNotAPack(json: String) {
        #expect(throws: WordPackError.self) { try WordPackReader.read(Self.data(json)) }
    }

    @Test func refusesAPackThatChoosesItsOwnFileName() {
        let json = Self.valid.replacingOccurrences(of: "travel-basics", with: "../../words")

        #expect(throws: WordPackError.unsafeId) { try WordPackReader.read(Self.data(json)) }
    }

    @Test func refusesAPackWithNoName() {
        let json = Self.valid.replacingOccurrences(of: "Travel basics", with: "   ")

        #expect(throws: WordPackError.malformed("the pack has no name")) {
            try WordPackReader.read(Self.data(json))
        }
    }

    @Test func refusesAPayloadOverTheLimit() {
        // Padding inside the description, so the file stays valid JSON.
        let padding = String(repeating: "x", count: PackLimits.maxPayloadBytes + 1)
        let json = Self.valid.replacingOccurrences(of: "Everyday phrases", with: padding)

        #expect(throws: WordPackError.tooLarge) { try WordPackReader.read(Self.data(json)) }
    }

    @Test func refusesTooManyWords() {
        let entries = (0...PackLimits.maxWords)
            .map { #"{ "original": "w\#($0)", "translation": "t\#($0)" }"# }
            .joined(separator: ",")

        #expect(throws: WordPackError.tooLarge) {
            try WordPackReader.read(Self.data(#"{ "id": "big", "name": "Big", "words": [\#(entries)] }"#))
        }
    }

    @Test func refusesAFieldTooLongToBeAWord() {
        let essay = String(repeating: "a", count: PackLimits.maxFieldLength + 1)
        let json = #"{ "id": "x", "name": "X", "words": [{ "original": "\#(essay)", "translation": "t" }] }"#

        #expect(throws: WordPackError.self) { try WordPackReader.read(Self.data(json)) }
    }

    @Test func skipsBlankEntriesButKeepsTheRest() throws {
        let json = """
            { "id": "x", "name": "X", "words": [
              { "original": "  ", "translation": "empty" },
              { "original": "keep", "translation": "" },
              { "original": "airport", "translation": "aeropuerto" }] }
            """

        let pack = try WordPackReader.read(Self.data(json))

        #expect(pack.words.count == 1)
        #expect(pack.words[0].original == "airport")
    }

    @Test func refusesAPackWhereNothingUsableIsLeft() {
        let json = #"{ "id": "x", "name": "X", "words": [{ "original": " ", "translation": " " }] }"#

        #expect(throws: WordPackError.empty) { try WordPackReader.read(Self.data(json)) }
    }

    @Test func refusesAPackWithNoWordsAtAll() {
        #expect(throws: WordPackError.empty) {
            try WordPackReader.read(Self.data(#"{ "id": "x", "name": "X", "words": [] }"#))
        }
    }

    @Test func foldsControlCharactersThatWouldReachANotification() throws {
        let json = """
            { "id": "x", "name": "X", "words": [
              { "original": "two\\nlines", "translation": "\\ttabbed " }] }
            """

        let word = try WordPackReader.read(Self.data(json)).words[0]

        #expect(word.original == "two lines")
        #expect(word.translation == "tabbed")
    }

    @Test func truncatesTheNameRatherThanRefusingThePack() throws {
        // Display-only, and the reader cannot expect the user to fix someone else's file.
        let json = Self.valid.replacingOccurrences(
            of: "Travel basics",
            with: String(repeating: "n", count: 200))

        #expect(try WordPackReader.read(Self.data(json)).name.count == PackLimits.maxNameLength)
    }
}
